### PR TITLE
fix bug when set vmx_cr0/4

### PR DESCRIPTION
**Pull request by lifeix**
When setting some bits of vmx cr0/4, the guest would trap out. If these bits
want to enable/disable the paging or modify paging access right, we should
invalidate cached translation information. This serial is for this.

Tracked-On: #1379
v1->v2:
return bool for guest paging related APIs which added by this patch.

### CODE REVIEW

**Comment by HuiHuangShi** [comment]
1377<br>3 new violations to the coding guideline detected. <br><br>1.filename:/hypervisor/include/arch/x86/guest/vcpu.h&nbsp;&nbsp;function:is_long_mode&nbsp;&nbsp;offset:290:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Static procedure is not explicitly called in code analysed. : is_long_mode<br><br>2.filename:/hypervisor/arch/x86/vmx.c&nbsp;&nbsp;function:load_pdptrs&nbsp;&nbsp;offset:299:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Casting operation on a pointer. : (void* to unsigned long*): ( uint64_t * ) gpa2hva ( vcpu -> vm , guest_cr3 )<br><br>3.filename:/hypervisor/arch/x86/vmx.c&nbsp;&nbsp;function:load_pdptrs&nbsp;&nbsp;offset:299:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Casting operation to a pointer. : (void* to unsigned long*): ( uint64_t * ) gpa2hva ( vcpu -> vm , guest_cr3 )<br><br>

File: hypervisor/arch/x86/vmx.c [hunk 1]
```diff
@@ -481,13 +506,24 @@ static bool is_cr4_write_valid(uint64_t cr4)
...
+	uint64_t old_cr4 = vcpu_get_cr4(vcpu);
 
-	if (!is_cr4_write_valid(cr4)) {
+	if (!is_cr4_write_valid(vcpu, cr4)) {
 		pr_dbg("Invalid cr4 write operation from guest");
 		vcpu_inject_gp(vcpu, 0U);
 		return;
 	}
 
+	if ((cr4 ^ old_cr4) & (CR4_PGE | CR4_PSE | CR4_PAE |
```
Comment (dongyaozu) [inline hypervisor/arch/x86/vmx.c]: for conditions, please use " a == b" or "a!= b" etc.. Please fix in all of the files.

File: hypervisor/arch/x86/vmx.c [hunk 2]
```diff
@@ -481,13 +506,24 @@ static bool is_cr4_write_valid(uint64_t cr4)
...
+			CR4_SMEP | CR4_SMAP | CR4_PKE)) {
+		if ((cr4 & CR4_PAE) && is_paging_enabled(vcpu) &&
```
Comment (dongyaozu) [inline hypervisor/arch/x86/vmx.c]: ditto

**Comment by HuiHuangShi** [comment]
1377<br>3 new violations to the coding guideline detected. <br><br>1.filename:/hypervisor/include/arch/x86/guest/vcpu.h&nbsp;&nbsp;function:is_long_mode&nbsp;&nbsp;offset:290:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Static procedure is not explicitly called in code analysed. : is_long_mode<br><br>2.filename:/hypervisor/arch/x86/vmx.c&nbsp;&nbsp;function:load_pdptrs&nbsp;&nbsp;offset:299:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Casting operation on a pointer. : (void* to unsigned long*): ( uint64_t * ) gpa2hva ( vcpu -> vm , guest_cr3 )<br><br>3.filename:/hypervisor/arch/x86/vmx.c&nbsp;&nbsp;function:load_pdptrs&nbsp;&nbsp;offset:299:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Casting operation to a pointer. : (void* to unsigned long*): ( uint64_t * ) gpa2hva ( vcpu -> vm , guest_cr3 )<br><br>

File: hypervisor/arch/x86/vmx.c (hunk 1 above)
Comment (lifeix) [reply hypervisor/arch/x86/vmx.c]: done

File: hypervisor/arch/x86/vmx.c (hunk 2 above)
Comment (lifeix) [reply hypervisor/arch/x86/vmx.c]: done

File: hypervisor/arch/x86/vmx.c [hunk 3]
```diff
@@ -360,37 +373,39 @@ void vmx_write_cr0(struct vcpu *vcpu, uint64_t cr0)
...
+			entry_ctrls = exec_vmread32(VMX_ENTRY_CONTROLS);
+			entry_ctrls |= VMX_ENTRY_CTLS_IA32E_MODE;
+			exec_vmwrite32(VMX_ENTRY_CONTROLS, entry_ctrls);
+
+			vcpu_set_efer(vcpu,
+				vcpu_get_efer(vcpu) | MSR_IA32_EFER_LMA_BIT);
+		} else if (is_pae(vcpu)) {
+			/* enabled PAE from paging disabled */
+			load_pdptrs(vcpu);
+		}
```
Comment (dongyaozu) [inline hypervisor/arch/x86/vmx.c]: we need end this with else {}.   This is also MISRAC requirement

File: hypervisor/arch/x86/vmx.c [hunk 4]
```diff
@@ -360,37 +373,39 @@ void vmx_write_cr0(struct vcpu *vcpu, uint64_t cr0)
...
+			/* Disable long mode */
+			pr_dbg("VMM: Disable long mode");
+			entry_ctrls = exec_vmread32(VMX_ENTRY_CONTROLS);
+			entry_ctrls &= ~VMX_ENTRY_CTLS_IA32E_MODE;
+			exec_vmwrite32(VMX_ENTRY_CONTROLS, entry_ctrls);
+
+			vcpu_set_efer(vcpu,
+				vcpu_get_efer(vcpu) & ~MSR_IA32_EFER_LMA_BIT);
+		}
 	}
```
Comment (dongyaozu) [inline hypervisor/arch/x86/vmx.c]: ditto

**Comment by HuiHuangShi** [comment]
1377<br>3 new potential violations to the coding guideline detected. Please help check. Thanks!  <br><br>1.filename:/hypervisor/include/arch/x86/guest/vcpu.h&nbsp;&nbsp;function:is_long_mode&nbsp;&nbsp;offset:290:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Static procedure is not explicitly called in code analysed. : is_long_mode<br><br>2.filename:/hypervisor/arch/x86/vmx.c&nbsp;&nbsp;function:load_pdptrs&nbsp;&nbsp;offset:299:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Casting operation on a pointer. : (void* to unsigned long*): ( uint64_t * ) gpa2hva ( vcpu -> vm , guest_cr3 )<br><br>3.filename:/hypervisor/arch/x86/vmx.c&nbsp;&nbsp;function:load_pdptrs&nbsp;&nbsp;offset:299:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Casting operation to a pointer. : (void* to unsigned long*): ( uint64_t * ) gpa2hva ( vcpu -> vm , guest_cr3 )<br><br>

File: hypervisor/arch/x86/vmx.c (hunk 3 above)
Comment (lifeix) [reply hypervisor/arch/x86/vmx.c]: done

File: hypervisor/arch/x86/vmx.c (hunk 4 above)
Comment (lifeix) [reply hypervisor/arch/x86/vmx.c]: done

**Comment by HuiHuangShi** [comment]
1377<br>4 new potential violations to the coding guideline detected. Please help check. Thanks!  <br><br>1.filename:/hypervisor/include/arch/x86/guest/vcpu.h&nbsp;&nbsp;function:is_long_mode&nbsp;&nbsp;offset:290:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Static procedure is not explicitly called in code analysed. : is_long_mode<br><br>2.filename:/hypervisor/arch/x86/vmx.c&nbsp;&nbsp;function:load_pdptrs&nbsp;&nbsp;offset:299:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Casting operation on a pointer. : (void* to unsigned long*): ( uint64_t * ) gpa2hva ( vcpu -> vm , guest_cr3 )<br><br>3.filename:/hypervisor/arch/x86/vmx.c&nbsp;&nbsp;function:load_pdptrs&nbsp;&nbsp;offset:299:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Casting operation to a pointer. : (void* to unsigned long*): ( uint64_t * ) gpa2hva ( vcpu -> vm , guest_cr3 )<br><br>4.filename:/hypervisor/arch/x86/vmx.c&nbsp;&nbsp;function:vmx_write_cr0&nbsp;&nbsp;offset:391:<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reason:Empty else clause following else if.<br><br>

**Comment by haoweiah** [comment]
Ready to merge

**Pull request merged**